### ISSUE TITLE
Compilation error with gcc 4.9.4: "error: unknown option after ‘#pragma GCC diagnostic’ kind"

### ISSUE THREAD

**Issue by olifre** [OpenRCT2/OpenRCT2#6030]
**OS:** Gentoo Linux x86_64
**Version:** gcc 4.9.4
**Commit/Build:** v0.1.0

Compiling, I get the following errors:
```
In file included from /var/tmp/portage/games-simulation/openrct2-0.1.0/work/OpenRCT2-0.1.0/src/openrct2/drawing/X8DrawingEngine.cpp:27:0:
/var/tmp/portage/games-simulation/openrct2-0.1.0/work/OpenRCT2-0.1.0/src/openrct2/drawing/X8DrawingEngine.h:63:32: error: unknown option after ‘#pragma GCC diagnostic’ kind [-Werror=pragmas]
 #pragma GCC diagnostic ignored "-Wsuggest-final-types"
                                ^
/var/tmp/portage/games-simulation/openrct2-0.1.0/work/OpenRCT2-0.1.0/src/openrct2/drawing/X8DrawingEngine.cpp:137:32: error: unknown option after ‘#pragma GCC diagnostic’ kind [-Werror=pragmas]
 #pragma GCC diagnostic ignored "-Wsuggest-final-methods"
                                ^
cc1plus: all warnings being treated as errors
```

These options are not supported by gcc 4.9.4 yet...

**Comment by janisozaur** [comment]
GCC 4.9 is a very old compiler, I'm inclined to close this PR with a wontfix resolution.

Other than these pragmas, does it compile?

**Comment by Gymnasiast** [comment]
I agree with @janisozaur on this. Please upgrade to a GCC from the 6.x or 7.x series.

**Comment by olifre** [comment]
I tend to disagree. Various reasons:
- Removing the pragmas in these two files, it compiles perfectly fine with gcc 4.9.4 (!). Since your main `CMakeLists.txt` checks whether the compiler supports these warning-flags, it would only be correct to also only have this pragmas in if the compiler actually supports them. 
- gcc 6 and gcc 7 are not yet used on *any* distribution as system compilers as far as I know. Please correct me if I am wrong. 
- If you really stay on this decision, please (at least) update the build requirements to include that. Then you would be one of the first packages in Gentoo explicitly requiring something newer than gcc-4. Also, the Gentoo package then needs fixing to take this special requirement into account... 

Let me point out a few additional things you might not be aware of: 
- Most enterprise-grade distributions are still with gcc 4 (for good reasons - gcc 5 shows ABI inconsistencies, see e.g. [https://reviews.llvm.org/D18035](url) and newer gccs are still in testing, since not all packages compile with them). A notable example for such a distribution which is used by many is the fresh openSUSE 42.2, so staying on this decision means you lose support for openSUSE right now - just for two pragmas. 
- Mixing a gcc-4 compiled system with gcc-5 (or 6 or 7) compiled applications is not supported and might lead to erratic / undefined behaviour (potentially only at runtime) due to the ABI changes. 

Please let me know what you think - at least, I'd say that either the build requirements and the Gentoo package need fixing, or alternatively, the use of those pragmas in the two mentioned files needs to be adapted. So for sure, something needs to be done here - please re-open and take your decision.

**Comment by Gymnasiast** [comment]
> gcc 6 and gcc 7 are not yet used on any distribution as system compilers as far as I know. Please correct me if I am wrong.

Ubuntu has used GCC 6 since 16.10. Debian Stretch also uses 6.

> A notable example for such a distribution which is used by many is the fresh openSUSE 42.2

Strangely, we have had no such complaints from the openSUSE maintainer. openSUSE does have GCC 5 available.


I guess it's mostly a problem for Gentoo as it compiles everything on the user's machine. On Ubuntu , one can upgrade GCC with no problem as it only affects stuff the user compiles themselves anyway.

**Comment by Gymnasiast** [comment]
Oh, by the way, we will accept a PR that enables compilation on GCC 4.9.4, provided it isn't intrusive - if it's indeed only #ifdef-ing two pragmas, that would count as non-intrusive.

**Comment by janisozaur** [comment]
Personally I think `#ifdef`-ing out these pragmas is unintrusive enough to be merged, if you were to submit a PR – if that's the only thing preventing the project from being able to compile.

Keep in mind that even without those pragmas we won't consider [EOL'd GCC 4.9](https://gcc.gnu.org/gcc-4.9/) as something that we explicitly support. On top of that, we want to move forward with used C++ version, so even GCC5 may be on its way out and we won't be bothered.

> gcc 6 and gcc 7 are not yet used on any distribution as system compilers as far as I know.

See https://fedoraproject.org/wiki/Fedora_26_Mass_Rebuild and https://fedoraproject.org/wiki/Changes/GCC7 for F26, https://fedoraproject.org/wiki/Changes/GCC6 for F24. Debian Stretch ships with GCC6 by default, and sure enough, the logs from building `linux` package show it's being used: https://buildd.debian.org/status/fetch.php?pkg=linux&arch=amd64&ver=4.11.6-1%7Ebpo9%2B1&stamp=1500040080&raw=0. Now you know at least two distributions using GCC6+ as their system compilers.

> Most enterprise-grade distributions are still with gcc 4

I can give you that, RHEL7 (and thus CentOS7), SLES12 and Oracle 7 (these are all that come to my mind when thinking about "enterprise") ship with GCC 4 by default, though I'm sure newer toolchains are available. That said we do not target _enterprise Linux_ at all, so this whole thing is not an issue in my point of view.

**TL;DR:** Submit a PR to `#ifdef`-out the pragmas and it has high chances of getting merged. We _will_ drop support for GCC4.9 as soon as we start using features it doesn't have.

**Comment by olifre** [comment]
> Strangely, we have had no such complaints from the openSUSE maintainer. openSUSE does have GCC 5 available.

Ok - I will check how the compilation is done there, before I look into opening a PR. They certainly will *not* mix gcc versions in the compilation of their packages, and their system is compiled with gcc 4 still. 


> I guess it's mostly a problem for Gentoo as it compiles everything on the user's machine. On Ubuntu , one can upgrade GCC with no problem as it only affects stuff the user compiles themselves anyway.

That's untrue: Updating the compiler and using it to compile only "some" parts of the system against system libraries compiled with a different gcc version opens up the full potential for ABI collisions (especially due to mangling bugs in gcc 5). I'm in fact planning to update my Gentoo systems to newer gccs, but since gcc 5 is inherently borked (mangling inconsistencies), I would tend to wait until gcc 6 is stable. Up to now, all packages in Gentoo's tree build with gcc 4 just fine. 


> TL;DR: Submit a PR to #ifdef-out the pragmas and it has high chances of getting merged. We will drop support for GCC4.9 as soon as we start using features it doesn't have.

Will do, many thanks!

**Comment by janisozaur** [comment]
1. ifdef is not the only workaround, having cmake detect gcc<5 and setting it to ignore such pragmas may be viable option. There might be others too, so please consider them carefully.
2. I think there are two versions of openSUSE: tumbleweed and leap, the package may only be supported on the former.
3. We've been wanting to use C++17 for a while now, so please take note of #6034 
4. As far as I know, ABI breakage came in GCC5 due to C++11. If your system already is built using GCC5, you should be relatively safe wrt to that issue. Going by your logic, one would have to rebuild whole system even for the most minutiae updates of the compiler, which obviously does not happen.

As a side note: I always considered gentoo to be rolling and bleeding edge, I'm surprised to find out that's not the case.

**Comment by olifre** [comment]
> ifdef is not the only workaround, having cmake detect gcc<5 and setting it to ignore such pragmas may be viable option. There might be others too, so please consider them carefully.

I agree. ifdef would by my last resort, I'll have a look to this later today. 



> I think there are two versions of openSUSE: tumbleweed and leap, the package may only be supported on the former.

I checked. There is *no* official package for openSUSE. There is a package in the "games" repository and some user building it, but these packages indeed use gcc 6 for building - which means it will never become an official part of 42.2 at least, since this mixing may lead to runtime errors. In case of gcc-4 / gcc-5 mixing, it does in many cases, the situation has improved a bit with the many fixes that went into gcc-6 (but upstream does not guarantee this works!). 


> As far as I know, ABI breakage came in GCC5 due to C++11. If your system already is built using GCC5, you should be relatively safe wrt to that issue. Going by your logic, one would have to rebuild whole system even for the most minutiae updates of the compiler, which obviously does not happen.

There have been ABI breakages between gcc-3 and gcc-4, and again between gcc-4 and gcc-5. In these cases, it is indeed *required* to rebuild all packages compiled with gcc (in case of the 4 => 5 transition, only if c++11 features are used). This is what all distributions I know are also doing. They have to, or will encounter compile-time errors or run-time undefined behaviour. 

Additionally, gcc-5 features some mangling inconsistencies, which is why clang can only be used in conjunction with it if you use a pretty recent clang version, where they added support for the broken mangling scheme. 

In between gcc 5, 6 and 7, compatbility is fine, so here recompilation of the full system is not necessary (albeit one may profit from better optimizations, of course). So yes, I agree, on a gcc 5 system, mixing newer gccs is (currently) working - of course, corner cases may exist. 



> As a side note: I always considered gentoo to be rolling and bleeding edge, I'm surprised to find out that's not the case.

They are careful in their selection of gcc versions. By now, gcc 4 and gcc 5 are called "stable", even though gcc 5 has known issues. gcc 6 is marked "unstable" (many packages break), gcc 7 is masked for testing (even more package break). 
Since in fact one needs to recompile almost everything based on C++ when switching from gcc 4 to gcc 5, and gcc 5 was marked stable not too long ago, most "production" Gentoo systems you find will still use gcc 4. In fact, I am already operating one Gentoo machine with gcc 6, and it is running well. Still, many packages that broke especially with gcc 5 made me wary of updating at the earliest time possible for a production system (and Gentoo allows you to choose how "bleeding edge" you really want to be, in a fine-grained manner).
